sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala

```diff
@@ -678,4 +679,45 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
       assert(e.contains("User specified schema not supported with `table`"))
     }
   }
+
+  test("SPARK-21144 fails when data schema and partition schema have duplicate columns") {
+    // Check CSV format
+    Seq((true, ("a", "a")), (false, ("aA", "Aa"))).foreach { case (caseSensitive, (c0, c1)) =>
+      withSQLConf(SQLConf.CASE_SENSITIVE.key -> caseSensitive.toString) {
+        withTempDir { src =>
+          Seq(1).toDF(c0).write.option("header", true).mode("overwrite").csv(s"$src/$c1=1")
+          val e = intercept[AnalysisException] {
+            spark.read.option("header", true).csv(src.toString)
+          }
+          assert(e.getMessage.contains("Found duplicate column(s) in the datasource: "))
+        }
+      }
+    }
+
+    // Check JSON format
+    Seq((true, ("a", "a")), (false, ("aA", "Aa"))).foreach { case (caseSensitive, (c0, c1)) =>
+      withSQLConf(SQLConf.CASE_SENSITIVE.key -> caseSensitive.toString) {
+        withTempDir { src =>
+          Seq(1).toDF(c0).write.mode("overwrite").json(s"$src/$c1=1")
```

Comment:
Hi, @maropu .
It seems we can simply merge `JSON` and `Parquet` test case into one by using `format(...)` instead of `json()` or `parquet()`.